tests/components/mqtt/test_init.py

```diff
@@ -1305,17 +1305,193 @@ def _callback_b(msg: ReceiveMessage) -> None:
     await hass.async_block_till_done()
     await mqtt.async_subscribe(hass, "test/state", _callback_b, qos=1)
     async_fire_mqtt_message(
-        hass, "test/state", "online"
-    )  # Simulate a (retained) message replaying
+        hass, "test/state", "online", qos=0, retain=False
+    )  # Simulate an other non retained message after the second subscription
     async_fire_time_changed(hass, utcnow() + timedelta(seconds=1))
     await hass.async_block_till_done()
     async_fire_time_changed(hass, utcnow() + timedelta(seconds=1))
     await hass.async_block_till_done()
+    # Both subscriptions should receive updates
     assert len(calls_a) == 1
     assert len(calls_b) == 1
     mqtt_client_mock.subscribe.assert_called()
 
 
+@patch("homeassistant.components.mqtt.client.INITIAL_SUBSCRIBE_COOLDOWN", 0.0)
+@patch("homeassistant.components.mqtt.client.DISCOVERY_COOLDOWN", 0.0)
+@patch("homeassistant.components.mqtt.client.SUBSCRIBE_COOLDOWN", 0.0)
+async def test_replaying_payload_same_topic(
+    hass: HomeAssistant,
+    mqtt_client_mock: MqttMockPahoClient,
+    mqtt_mock_entry_no_yaml_config: MqttMockHAClientGenerator,
+) -> None:
+    """Test replaying retained messages.
+
+    When subscribing to the same topic again, SUBSCRIBE must be sent to the broker again
+    for it to resend any retained messages for new subscriptions.
+    Retained messages must only be replayed for new subscriptions, except
+    when the MQTT client is reconnection.
+    """
+    mqtt_mock = await mqtt_mock_entry_no_yaml_config()
+
+    # Fake that the client is connected
+    mqtt_mock().connected = True
+
+    calls_a: list[ReceiveMessage] = []
+    calls_b: list[ReceiveMessage] = []
+
+    def _callback_a(msg: ReceiveMessage) -> None:
+        calls_a.append(msg)
+
+    def _callback_b(msg: ReceiveMessage) -> None:
+        calls_b.append(msg)
+
+    await mqtt.async_subscribe(hass, "test/state", _callback_a)
+    async_fire_mqtt_message(
+        hass, "test/state", "online", qos=0, retain=True
+    )  # Simulate a (retained) message played back
+    await hass.async_block_till_done()
+    assert len(calls_a) == 1
+    mqtt_client_mock.subscribe.assert_called()
+    calls_a = []
+    mqtt_client_mock.reset_mock()
+
+    await mqtt.async_subscribe(hass, "test/state", _callback_b)
+    async_fire_mqtt_message(
```

Comment:
Would be nice to have a retain=false message get sent after callback_b is subscribed, but before the retain=true message for callback_b is sent.   Which should should both len(calls_a) and len(calls_b) ==1

ie. this test proves that a subscription CAN receive a retain=false message before a retain=true, and that's fine and it should be processed, but that should not (and does not currently) alter the behavior for retain=true messages (ie. only the first one for any subscription is processed).

Can be skipped if you like, but it puts in place a test of this "edge case" so it's codified how it should work.

Reply:
I have updated test `tests/components/mqtt/test_init.py::test_debug_info_qos_retain`.
It shows such a case by asserting on the diagnostics feature.
But if you like I could add another test as well or update an existing.

Reply:
BTW I have added an extra test that asserts this, including a case where we unsubscribe and resubscribe again and want to see an initial retained payload.

Reply:
As mentioned in my other comment - my concern here is that, due to the debounce the MQTT subscription will be delayed, but the Subscribe class will exist.  And if the _callback_a subscription here receives a RETAIN=False message during the debounce period (where the Subscribe class for _callback_a and _callback_b both exist, but we have not yet sent the subscribe for _callback_b to MQTT), both will receive the RETAIN=False message due to how we handle messages received.

It's arguable that in this case, the _callback_b should NOT be sent the RETAIN=0 message until it has received a RETAIN=1 message.  But that's not currently the case.

Reply:
I does not matter what message comes in at the moment it is subscribed, as I explained already, if our subscribe happens before an actual retained message is published, the message will not have retain flag. Further, the retain flag is not used for anything else then diagnostics or debug logging. I hope you get what I mean. The only thing we need to do is filter the retained messages if we did not receive a messages before on that topic for that subscription.

Reply:
The RETAIN flag is not for diagnostics.  the retain flag is for 'this is the full current state' - everything after could be deltas (or could be a full state, you won't know).

And I'm not worried about the RETAIN message being published to MQTT after our subscription.  I am worried about us receiving a RETAIN=0 message between when created the Subscribe class, and when we actually sent the subscription (due to debounce).
ie. there will still be 2 RETAIN=1 messages coming, and 1 RETAIN=0 message between them, but after the first RETAIN=1 and before the RETAIN=0 message the Subscribe class is created, then after the RETAIN=0 message is received, the subscribe message for that Subscribe class is sent to MQTT (which triggers the second RETAIN=1).

Reply:
> The RETAIN flag is not for diagnostics. the retain flag is for 'this is the full current state' - everything after could be deltas (or could be a full state, you won't know).
That is not True, the retain flag does not reflect the current state. The current state is always in the past. Even more, it can even be that it is not the last state (I already explained this). It only indicates there was a retained payload at the broker some time ago. Test for your self (you can use the MQTT diagnostic tools or any tool you like):
1. Subscribe to a topic (so we monitor it)
2. Publish a retained value to the topic "value 1" -> You should receive "value 1" without retain flag.
3. Publish a non retained value to the topic "value 2" -> You should receive "value 2"without retain flag.
4. Now unsubscribe and subscribe (or resubscribe) what ever you like -> You should receive "value 1" with retain flag, as this is the last retained payload.

Reply:
> ie. there will still be 2 RETAIN=1 messages coming, and 1 RETAIN=0 message between them, but after the first RETAIN=1 and before the RETAIN=0 message the Subscribe class is created, then after the RETAIN=0 message is received, the subscribe message for that Subscribe class is sent to MQTT (which triggers the second RETAIN=1).

If we get a retained publish after we subscribe it will be treated as a normal update (no retain flag), if the resubscribe was not executed yet and a message arrive, then that message will be the last message received. We do not known if that message was retained, after that when the PAHO client resubscribes the topic, it will replay the last retained value (that could be the last value, be we can not be sure about that). We can safely ignore the retained payload as we already got a valid last update from the message before resubscribing. If there is a longer time between publishes, with a retained message we make sure the last retained value will be served immediately.

Reply:
I understand that - but you're being disingenuous.  The topic doesn't have just 'value 1' or 'value 2'.  It has a complex JSON object.

A RETAIN=0 message (from the publisher) might only update ONE FIELD in that complex json (which may be considered 'a minor update', or even worse, could just be some advisory information on that topic (see where I sent an error message, rather than any real data).

The RETAIN=1 message is almost guaranteed to be the full state (it doesn't have to be if you have an insane publisher, but it should be).

As an example, I have MQTT devices - that device may have 25 different sensors on it.  the client may sent a RETAIN=0 with ONLY the individual sensor that updated on that device, rather than sending the ENTIRE state of the device.  And then periodically (or if the change is great enough) send a RETAIN=1 with the whole device state.

This could happen, say, if a device is powering up.  And it sends a RETAIN=0 message for current_power_usage frequently, but once it gets to it's final state, it sends a RETAIN=1, so that any new subscribers will now get this final state (they will get this as RETAIN=0, but it was sent with RETAIN=1).  But it neatly avoided sending 25 sensors worth of data during the power ramp up stage (because it doesn't matter TOO much if that value is wrong for a subscriber that subscribed during ramp up).

It could also be that this sensor is minor and changes frequently, and doesn't matter if it's wrong (eg. A/C frequency), so it's sent as a RETAIN=0 message alone.  But it's not worth sending out a full snapshot when the A/C frequency changes unless it changes drastically (ie. there is no guarantee of when a full snapshot will be upcoming).

If we ignore the RETAIN=1 message, because we got a RETAIN=0 message, which the current code does (because of the debounce delay), then that device will not have any clue of the state of the other 24 sensors, until the publisher decides to publish the entire state again - which, who knows when that will be.

Note: I am always explicitly talking about 2 subscriptions that weren't done at the same time.  With the debounce causing a delay between the Subscribe class being created, and the MQTT SUBSCRIBE being sent.  Not talking about any unsubscribes.

Reply:
I'll respond later. MQTT is a very simple protocol. It does not care about json fields. Event the slightest update means the whole state must be resent. For MQTT we talk about payloads, the example I gave demonstrates how retention works.
I have no idea what you mean with full state, every publish is the full state for a topic.
In my environment I do not use retain at all, so it takes a few seconds before my devices pop up, and that is fine in may cases. If a device has many sensors, it can either publish a full state of all values or use multiple topics for its sensors.
But we get of topic for this PR. Back o the top of this chain, I updated the test and that is it.

Reply:
MQTT doesn't require the full state be sent with every message.  MQTT doesn't even know it's dealing with state, for all it knows, you could pass a message along that has nothing to do with the previous messages.

See my code in the stand-alone comment, but I believe it's perfectly valid and reasonable to have messages that only partially update state.  Just because everything you use currently updates the entire state of an object with every message, doesn't mean it's required.  HA allows you to update certain fields/entities on a device and any you don't update, will be left as they were.